x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -275,221 +284,109 @@ public void checkCompatibility(MappedFieldType otherFT, List<String> conflicts)
+            BooleanQuery.Builder builder = new BooleanQuery.Builder();
+            assert ngramAutomata.length == automata.length;
+            for (int i=0; i<automata.length;i++) {
+                Query ngramQuery = automatonToQuery(ngramFuzzyPattern, searchTerm, ngramAutomata[i].automaton, automata[i].automaton);
+                builder.add(ngramQuery, Occur.SHOULD);                                
             }
-            // TODO we can be smarter about pruning here. e.g.
-            // * Avoid wildcard queries if there are sufficient numbers of other terms that are full 3grams that are cheaper term queries
-            // * We can select terms on their scarcity rather than even spreads across the search string.
+            return builder.build();
+        }
 
-            return bqBuilder.build();
+        protected Query automatonToQuery(String value, Automaton automaton) {
+            return automatonToQuery(value,  value, automaton, automaton);
         }
 
-        private void addClause(String token, BooleanQuery.Builder bqBuilder) {
-            assert token.codePointCount(0, token.length()) <= NGRAM_SIZE;
-            if (token.codePointCount(0, token.length()) == NGRAM_SIZE) {
-                TermQuery tq = new TermQuery(new Term(name(), token));
-                bqBuilder.add(new BooleanClause(tq, Occur.MUST));
+        // The ngram and dv values can differ - null chars added to string beginning and ends in ngram index and used by
+        // wildcard query expressions.
+        protected Query automatonToQuery(String ngramValue, String dvValue, Automaton ngramAutomaton, Automaton dvAutomaton) {
+            Query ngramApproximation;
+            Expression<String> expression = new NGramExtractor(NGRAM_SIZE, settings.getMaxExpand(), settings.getMaxStatesTraced(),
+                    settings.getMaxNgramsExtracted()).extract(ngramAutomaton).simplify();
+            if (expression.alwaysTrue()) {
+                if (settings.getRejectUnaccelerated()) {
+                    throw new UnableToAccelerateRegexException(ngramValue, NGRAM_SIZE, name());
+                }
+                ngramApproximation = new MatchAllDocsQuery();
+            } else if (expression.alwaysFalse()) {
+                // Not sure what regex expression would cause this branch but cover it anyway.
+                ngramApproximation = new MatchNoDocsQuery();
             } else {
-                WildcardQuery wq = new WildcardQuery(new Term(name(), token + "*"));
-                wq.setRewriteMethod(MultiTermQuery.CONSTANT_SCORE_REWRITE);
-                bqBuilder.add(new BooleanClause(wq, Occur.MUST));
+                ngramApproximation = expression.transform(new ExpressionToFilterTransformer(name()));
             }
+            //TODO - skip every other ngram in long strings
+            //TODO - don't generate terms not in index.
+            //TODO - prefer rare ngrams when pruning
```

Comment:
These are all *maybe* TODOs. All cool things.

Reply:
I wanted to leave a reminder somewhere. @jpountz and @jimczi had a number of optimisations in mind and I felt the first iteration would be to have a test framework that shows we do the same thing as keyword field. If we get that foundation in place we can iterate on optimisations knowing we haven't introduced any false negatives in the acceleration logic.